[nnc] Add hardsigmoid

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59405 Fix symbolic derivative of hardswish
* **#59069 [nnc] Add hardsigmoid**

Differential Revision: [D28738166](https://our.internmc.facebook.com/intern/diff/D28738166)

bertmaher committed:
[nnc] Add hardsigmoid

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28738166).

eellison reviewed (approved):
looks good! if you add symbolic script support it will also speed up training and not just inference :)

bertmaher commented:
> looks good! if you add symbolic script support it will also speed up training and not just inference :)

Nice!  Do we have a test harness for backward where I can add hardsigmoid to make sure I've done my differentiation correctly?

eellison commented:
Yep, you can follow as https://github.com/pytorch/pytorch/pull/58773/files

bertmaher committed:
Update on "[nnc] Add hardsigmoid"


Differential Revision: [D28738166](https://our.internmc.facebook.com/intern/diff/D28738166)

[ghstack-poisoned]

desertfire commented on torch/csrc/jit/runtime/symbolic_script.cpp:
```diff
@@ -910,6 +910,12 @@ const std::vector<std::string> functions = {
+        def hardsigmoid(self):
+            result = torch.hardsigmoid(self)
+            def backward(grad_output):
+                return torch.where((result > -3.) & (result < 3.), 0., grad_output * (1.0 / 6.0))
```
I think you need to swap the if and else part.

desertfire reviewed (changes requested):
symbolic_script.cpp needs to updated.

bertmaher commented:
Thank you.  I was misled by the order of args of Vec:vblend, and didn't think hard enough about the math :-(.  It's disturbing that no tests caught this?

bertmaher committed:
Update on "[nnc] Add hardsigmoid"


Differential Revision: [D28738166](https://our.internmc.facebook.com/intern/diff/D28738166)

[ghstack-poisoned]

eellison commented:
yea, that is not good... I think we need to add more warmup runs to trigger the path

Krovatkin commented:
@bertmaher could you try adding your test here : https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/jit_metaprogramming_utils.py I wonder if its the integration of autodiff testing into OpInfo or our actual autodiff testing that's borked.

bertmaher committed:
Update on "[nnc] Add hardsigmoid"


Differential Revision: [D28738166](https://our.internmc.facebook.com/intern/diff/D28738166)

[ghstack-poisoned]

bertmaher committed:
Update on "[nnc] Add hardsigmoid"


Differential Revision: [D28738166](https://our.internmc.facebook.com/intern/diff/D28738166)

[ghstack-poisoned]

bertmaher commented:
@bertmaher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28738166).
